src/ckb/channel.rs

```diff
@@ -1590,34 +1645,183 @@ impl ChannelActorState {
+        self.commitment_numbers.get_local()
     }
 
-    pub fn get_remote_channel_parameters(&self) -> &ChannelParametersOneParty {
-        self.remote_channel_parameters.as_ref().unwrap()
+    pub fn get_remote_commitment_number(&self) -> u64 {
+        self.commitment_numbers.get_remote()
+    }
+
+    pub fn increment_local_commitment_number(&mut self) {
+        self.commitment_numbers.increment_local();
+    }
+
+    pub fn increment_remote_commitment_number(&mut self) {
+        self.commitment_numbers.increment_remote();
     }
 
     pub fn get_current_commitment_number(&self, local: bool) -> u64 {
         if local {
-            self.local_commitment_number
+            self.get_local_commitment_number()
         } else {
-            self.remote_commitment_number
+            self.get_remote_commitment_number()
         }
     }
 
     pub fn get_next_commitment_number(&self, local: bool) -> u64 {
         self.get_current_commitment_number(local) + 1
     }
 
+    pub fn get_next_offering_tlc_id(&self) -> u64 {
+        self.tlc_ids.get_next_offering()
+    }
+
+    pub fn get_next_received_tlc_id(&self) -> u64 {
+        self.tlc_ids.get_next_received()
+    }
+
+    pub fn increment_next_offering_tlc_id(&mut self) {
+        self.tlc_ids.increment_offering();
+    }
+
+    pub fn increment_next_received_tlc_id(&mut self) {
+        self.tlc_ids.increment_received();
+    }
+
+    pub fn get_offered_tlc(&self, tlc_id: u64) -> Option<&DetailedTLCInfo> {
+        self.tlcs.get(&TLCId::Offered(tlc_id))
+    }
+
+    pub fn get_received_tlc(&self, tlc_id: u64) -> Option<&DetailedTLCInfo> {
+        self.tlcs.get(&TLCId::Received(tlc_id))
+    }
+
+    pub fn insert_tlc(&mut self, tlc: TLC) -> Result<DetailedTLCInfo, ProcessingChannelError> {
+        if let Some(current) = self.tlcs.get(&tlc.id) {
+            if current.tlc == tlc {
+                debug!(
+                    "Repeated processing of AddTlcCommand with id {:?}: current tlc {:?}",
+                    tlc.id, current,
+                );
+                return Ok(current.clone());
+            } else {
+                return Err(ProcessingChannelError::InvalidParameter(format!(
+                        "Trying to insert different tlcs with identical id {:?}: current tlc {:?}, new tlc {:?}",
+                        tlc.id, current, tlc
+                    )));
+            }
+        };
+        debug!(
+            "Adding new tlc {:?} to channel {:?} with local balance {} and remote balance {}",
+            &tlc,
+            &self.get_id(),
+            self.to_local_amount,
+            self.to_remote_amount
+        );
+        let detailed_tlc = DetailedTLCInfo {
+            tlc,
+            committed_at: self.get_current_commitment_numbers(),
+            removal_info: None,
+        };
+        self.tlcs.insert(tlc.id, detailed_tlc);
+        if tlc.is_offered() {
+            self.to_local_amount -= tlc.amount;
```

Comment:
can we add an `blance_check` function, which will make sure `to_local_amount + to_remote_amount + pending tlcs amount == initial channel amount`, so that we can catch error whenever unexpected thing happened.

Reply:
Added in https://github.com/nervosnetwork/cfn-node/pull/72/commits/19d43491f9c5ccf660131c106060b9117adb09d3